kernel/imf2imf.1
.TH IMF2IMF 1 
.SH NAME
imf2imf \- manipulate Xconq image families
.SH SYNOPSIS
.B imf2imf
.I [ options ] ...
.SH DESCRIPTION
.PP
.B imf2imf
reads and writes
.I Xconq
image families.
[include some examples of usage]
.LP
Options:
.TP 5
.B \-i \idir\fP
sets the input directory.
.TP 5
.B \-o \idir\fP
sets the input directory.
.TP 5
.B \-p
enables progress display.
.TP 5
.B \--explode
writes out multiple imf files, one for each image family.
.TP 5
.B \--help
displays help information.
.SH SEE ALSO
.BR xconq (6), x2imf (1), imf2x (1)
.SH AUTHOR
Stan Shebs (shebs@cygnus.com)


NEWS
XCONQ RELEASE NOTES

*** 7.5.0

This has not been released yet.
Current estimates put it at 1Q of 2005 sometime.
There are lots and lots of bugfixes and new features.

*** 7.4.1

The Windows port has not been built, and probably won't build without
some work, because of networking changes in shared Windows/Unix code.

Made cconq build again.

*** 7.4.0

Hans Ronne contributed a new Tolkien-inspired game "Lord of the Rings"
(lord-rings.g), and its supporting modules 3rd-age.g and t-midearth.g.

Keir Novik's "Carrier Battles" (tailhook.g), a detailed game of WWII
naval combat, and "Coral Sea", a Coral Sea scenario for it, are now
part of the game library.

Erik Sigra and Michael Burschik contributed better random name
generators, for Swedish and German place names respectively.

Standard flags now include 16-pixel-wide versions, and the 8-pixel
version use more colors.

The program tracks views of units using individual objects, which
enables more accurate display of multiple enemy units stacked at a
single location.

Networked games can be started from a mix of interactive and
command-line invocations.  The game module will not be downloaded
through the network connection, which makes startup faster.

Some unused and un-useful GDL constructs are gone, such as extension
properties, program-version, and image palettes.

The program supplies more feedback on why actions and tasks fail.

There is a "new" AI type named oplayer, which is just the previous
version of the AI.  This is mainly for testing that the AI is actually
getting smarter rather than dumber.

There is now more control available for the behavior of independent
units, for instance, you can set whether they will research and build
other units.

Sides now get colors instead of numbers if they have no emblems.

Games can use unit images that fill an entire hex, where previously
they were limited to a smaller bounding box within the hex.  This
is controlled by a new GDL variable "bigicons".

The tcl/tk interface now has a working isometric display mode.

The tcl/tk interface now lists only finished games by default,
and includes a button to list all games.

The Mac interface has many display improvements, from terrain
transitions to more floating windows.

The old Xt/Xaw interface has been removed, since it was no longer
being maintained.

There have been a host of minor improvements throughout the programs
and game libraries.

*** 7.3.3

Fix pathnames to tcl/tk libraries.

Fix Mac bugs: 1-bit images, loading of emblems from saved games, large
map handling, removed variants.

Add better solid colors in advterr.g.

*** 7.3.2

Fix crash with no-indepside games on Unix.

Add terrain images for advances.g.

*** 7.3.1

In Unix/Windows player setup, make separate buttons to add/remove
indepside AI and to configure indepside behavior.

Add --enable-alternate-scoresdir configure option.

Update config.guess and config.sub.

*** 7.3.0

This release accumulates over two years of changes throughout the
program.  In general, the emphasis has been on improving the program's
usability, making the AI smarter, improving existing games, and
extending to Civ-type games, rather than adding random new games to
the library.

The Unix user interface has been rewritten to use tcl/tk.  It includes
a full set of menus, resizeable panes in the map windows, buttons for
common operations, a mouseover display, and many additional map
display options.  There is also a full set of dialogs for setting up a
game, plus a chat window to facilitate setting up networked games.

There is now a port to Windows, using the tcl/tk-based interface,
which means no more requirement to run an X server.  The Windows port
still has bugs though.

The Mac interface now has floating windows for most auxiliary windows,
city and research dialogs, and more display controls, including
player controls over most colors and imagery used.  Selected units
can now blink rather than being surrounded with a box.

Unix networking support now allows for more than two players in a
game, and the game setup dialogs are synchronized, so for instance
clicking on a variant checkbox causes all players' checkboxes to
change.

Hans Ronne added the game "Ancient Near East" (anc-near-east.g),
similar to Civilization but with more detail, such as different kinds
of wheat to discover.  The included map of the Near East is
spectacularly large and detailed.

A Civilization II emulation (civ2.g) has been added.  The game works,
but happiness is not implemented, most of the city improvements have
no effect, nor do the Wonders.

Linn Stanton added an extended version of the standard game (lhs.g)
that includes radar, artillery, aaa, engineers, mines, and minefields.

Many new graphic images are available, particularly for terrain.
Xconq can load images directly from image files in standard formats
(although only GIF is available at present).

Players in the standard game get towns with names appropriate to
their chosen nationalities.

It is now possible to play independent units as if they were a regular
side, and (more usefully), it's possible to have an AI run the
independent units.  Watch out for the marauding barbarians!

A second AI type, the "iplayer", is available.  It is a minimal
AI that does basic tactical planning for individual units, but
does not attempt to coordinate them.

A set of commands, agreement-draft etc, are available for setting
up agreements.  (Agreement support is still incomplete though.)

New commands:
	"c-rate" sets conversion rates for materials (division of 
		trade into science/shields/luxuries in Civ, for instance).
	"collect" sets up a task to collect materials from terrain.
	"research" sets per-side research into advances.

There is a new GDL type "advance", to represent scientific or
technological advances.  Game designers can lay out a whole technology
tree, and either units or whole sides may do research to achieve
advances.


kernel/imf2imf.c
/* Manipulate Xconq image format files.
   Copyright (C) 1995, 1999 Stanley T. Shebs.

Xconq is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.  See the file COPYING.  */

#include "config.h"
#include "misc.h"
#include "lisp.h"
#include "imf.h"

char *indirname;

char *outdirname;

char imdirname[BUFSIZE];

int showprogress = 0;

int explode = 0;

char *xconqlib = "";

char spbuf[BUFSIZE];

char readerrbuf[1000];

static void usage(void);

int
main(int argc, char *argv[])
{
    int i, rslt;
    char *arg;
    FILE *ifp, *ofp, *fp;
    Obj *form;
    int startlineno = 0, endlineno = 0;

    init_lisp();

    if (argc == 1) usage();
    for (i = 1; i < argc; ++i) {
	arg = argv[i];
	if (strcmp(arg, "-i") == 0) {
	    if (i + 1 < argc) {
		indirname = argv[i + 1];
		/* Blast the arg because we'll be scanning the args again
		   and we want to ignore it then. */
		argv[i] = NULL;
		argv[i + 1] = NULL;
		++i;
	    } else {
		init_error("No input directory following -i");
	    }
	} else if (strcmp(arg, "-o") == 0) {
	    if (i + 1 < argc) {
		outdirname = argv[i + 1];
		/* Blast the arg because we'll be scanning the args again
		   and we want to ignore it then. */
		argv[i] = NULL;
		argv[i + 1] = NULL;
		++i;
	    } else {
		init_error("No output directory following -o");
	    }
	} else if (strcmp(arg, "-p") == 0) {
	    showprogress = 1;
	    argv[i] = NULL;
	} else if (strcmp(arg, "--help") == 0) {
	    usage();
	    argv[i] = NULL;
	} else if (strcmp(arg, "--explode") == 0) {
	    explode = 1;
	    argv[i] = NULL;
	}
    }
    /* Open and read an image directory file (and associated files)
       if supplied. */
    if (indirname != NULL) {
	ifp = fopen(indirname, "r");
	if (ifp != NULL) {
	    load_image_families(ifp, TRUE, NULL);
	} else {
	    run_warning("Couldn't open \"%s\", ignoring", indirname);
	}
    }
    /* Now append contents of random imf files. */
    for (i = 1; i < argc; ++i) {
	if (argv[i] != NULL) {
	    rslt = load_imf_file(argv[i], NULL);
	    if (!rslt)
	      run_warning("Couldn't open \"%s\", ignoring", argv[i]);
	}
    }
    sort_all_images();
    if (outdirname) {
	/* Write the image directory file. */
	sprintf(imdirname, "%s%s", outdirname, "imf.dir");
	ofp = fopen(imdirname, "w");
    } else {
	ofp = NULL;
    }
    for (i = 0; i < numimages; ++i) {
	if (showprogress)
	  fprintf(stderr, "; %s\n", images[i]->name);
	if (explode && outdirname) {
	    sprintf(spbuf, "%s%s", outdirname, images[i]->name);
	    fp = fopen(spbuf, "w");
	    if (fp != NULL) {
		write_imf(fp, images[i]);
		fclose(fp);
	    } else {
		fprintf(stderr, "could not open \"%s\"\n", spbuf);
	    }
	} else {
	    write_imf(stdout, images[i]);
	}
    }
    if (ofp != NULL) {
	fclose(ofp);
    }
    return 0;
}

static void
usage(void)
{
    fprintf(stderr,
	    "usage: imf2imf [ -i indir ] [ -o outdir ] [ -p ] [ files ... ]\n");
    exit(1);
}

void
close_displays(void)
{
}

int
write_entire_game_state(char *fname)
{
    return 0;
}

/* Need these so we can link in Lisp reader code. */

void
announce_read_progress()
{
}

void  
syntax_error(Obj *x, char *msg)
{
    sprintlisp(readerrbuf, x, BUFSIZE);
    init_warning("syntax error in `%s' - %s", readerrbuf, msg);
}

void
low_init_error(char *str)
{
    fprintf(stderr, "Error: %s.\n", str);
    fflush(stderr);
}

/* A warning just gets displayed, no other action is taken. */

void
low_init_warning(char *str)
{
    fprintf(stdout, "Warning: %s.\n", str);
    fflush(stdout);
}

/* A run error is fatal. */

void
low_run_error(char *str)
{
    fprintf(stderr, "Error: %s.\n", str);
    fflush(stderr);
    exit(1);
}

void
low_run_warning(char *str)
{
    fprintf(stdout, "Warning: %s.\n", str);
    fflush(stdout);
}

/* Fake definitions of unneeded routines called by lisp.c. */

int
keyword_code(char *str)
{
    run_warning("fake keyword_code being called");
    return 0;
}

/* Make the table so keyword lookup works. */

struct a_key {
    char *name;
} keywordtable[] = {

#undef  DEF_KWD
#define DEF_KWD(NAME,code)  { NAME },

#include "keyword.def"

    { NULL }
};

char *
keyword_name(enum keywords k)
{
    return keywordtable[k].name;
}

void
init_predefined_symbols(void)
{
}

int
lazy_bind(Obj *sym)
{
    run_warning("fake lazy_bind being called");
    return FALSE;
}

void
prealloc_debug(void)
{
}
